uniffi_core/src/lib.rs

```diff
@@ -124,13 +129,34 @@ macro_rules! assert_compatible_version {
 ///
 /// The `FfiConverter` trait defines how to pass values of a particular type back-and-forth over
 /// the uniffi generated FFI layer, both as standalone argument or return values, and as
-/// part of serialized compound data structures.
+/// part of serialized compound data structures. `FfiConverter` is mainly used in generated code.
+/// The goal is to make it easy for the code generator to name the correct FFI-related function for
+/// a given type.
 ///
 /// FfiConverter has a generic parameter, that's filled in with a type local to the UniFFI consumer crate.
 /// This allows us to work around the Rust orphan rules for remote types. See
 /// `https://mozilla.github.io/uniffi-rs/internals/lifting_and_lowering.html#code-generation-and-the-fficonverter-trait`
 /// for details.
 ///
+/// ## Scope
+///
+/// It could be argued that FfiConverter handles too many concerns and should be split into
+/// separate traits (like `FfiLiftAndLower`, `FfiSerialize`, `FfiReturn`).  However, there are good
+/// reasons to keep all the functionality in one trait.
+///
+/// The main reason is that splitting the traits requires fairly complex blanket implementations,
+/// for example `impl<UT, T> FfiReturn<UT> for T: FfiLiftAndLower<UT>`.  Writing these impls often
+/// means fighting with `rustc` over what counts as a conflicting implemmentation.  In fact, as of
+/// Rust 1.66, that previous example conflicts with `impl<UT> FfiReturn<UT> for ()`, since other
+/// crates can implement `impl FfiReturn<MyLocalType> for ()`. In general, this path gets
+/// complicated very quickly and that distracts from the logic that we want to define, which is
+/// fairly simple.
+///
+/// The main downside of having a single `FfiConverter` trait is that we need to implement it for
+/// types that only support some of the functionality.  For example, `Result<>` supports returning
+/// values, but not lifting/lowering/serializing them.  This is a bit weird, but works okay --
+/// especially since `FfiConverter` is rarely used outside of generated code.
```

Comment:
This is what I'm currently thinking, but I'm not 100% against splitting them up.  We would just need to split them up in a way where we don't have trait implementations dependent on other ones.  Maybe we could follow @jplatte's suggestion and split it up into `Lift`, `Lower`, and `Metadata`.  However, I just don't see a need to make the API super clean since the main consumer is our generate code.